File: 3-largest-prime-factor/3.hs
-- Find the largest prime factor of 600851475143

largestPrimeFactor :: (Integral a) => a -> a
largestPrimeFactor x = largestFactorAbove 3 x where
  largestFactorAbove start x
      | start^2 > x = x
      | (mod x start == 0) = largestFactorAbove (start + 2) (quot x start)
      | otherwise = largestFactorAbove (start + 2) x
                    
main = print (largestPrimeFactor 600851475143)

File: 3-largest-prime-factor/3.cpp
#include <iostream>

unsigned long largestPrimeFactor(unsigned long x);

int main() {
  std::cout << largestPrimeFactor(600851475143) << std::endl;
  return 0;
}

unsigned long largestPrimeFactor(unsigned long x) {
  unsigned long factor = 3;
  unsigned long largest = 3;
  while (x > largest) {
    if (x % factor == 0) {
      largest = factor;
      x /= factor;
    } else {
      factor += 2;
    }
  }

  return largest;
}
